plugin/plugin.go

```diff
@@ -0,0 +1,190 @@
+package plugin
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/sirupsen/logrus"
+)
+
+// Severity levels as constants for reuse and readability.
+const (
+	SeverityCritical = "critical"
+	SeverityWarning  = "warning"
+	SeverityInfo     = "info"
+)
+
+// Args provides plugin execution arguments.
+type Args struct {
+	Level             string `envconfig:"PLUGIN_LOG_LEVEL"`
+	RoutingKey        string `envconfig:"PLUGIN_ROUTING_KEY"`
+	IncidentSummary   string `envconfig:"PLUGIN_INCIDENT_SUMMARY"`
+	IncidentSource    string `envconfig:"PLUGIN_INCIDENT_SOURCE"`
+	IncidentSeverity  string `envconfig:"PLUGIN_INCIDENT_SEVERITY"`
+	DedupKey          string `envconfig:"PLUGIN_DEDUP_KEY"`
+	CreateChangeEvent bool   `envconfig:"PLUGIN_CREATE_CHANGE_EVENT"`
+	Resolve           bool   `envconfig:"PLUGIN_RESOLVE"`
+	JobStatus         string `envconfig:"PLUGIN_JOB_STATUS"`
+	CustomDetailsStr  string `envconfig:"PLUGIN_CUSTOM_DETAILS"` // Intermediate string to receive JSON
+	CustomDetails     map[string]interface{}
+}
+
+// PagerDutyClient defines the methods used from the PagerDuty API.
+type PagerDutyClient interface {
+	ManageEventWithContext(ctx context.Context, event *pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+	CreateChangeEventWithContext(ctx context.Context, event pagerduty.ChangeEvent) (*pagerduty.ChangeEventResponse, error)
+}
+
+// Exec executes the plugin.
+func Exec(ctx context.Context, client PagerDutyClient, args Args) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"PLUGIN_ROUTING_KEY":         args.RoutingKey,
+		"PLUGIN_INCIDENT_SUMMARY":    args.IncidentSummary,
+		"PLUGIN_INCIDENT_SOURCE":     args.IncidentSource,
+		"PLUGIN_INCIDENT_SEVERITY":   args.IncidentSeverity,
+		"PLUGIN_CREATE_CHANGE_EVENT": args.CreateChangeEvent,
+		"PLUGIN_JOB_STATUS":          args.JobStatus,
+	})
+
+	logger.Info("Starting plugin execution")
+
+	// Validate required fields
+	if args.RoutingKey == "" {
+		return errors.New("routingKey is required")
```

Comment:
```suggestion
		return errors.New("missing required parameter: routingKey")
```

make the error message clear for all the inputs